File: Code/js/Lesson_7_code/migrate.js
const fs = require('fs');
const path = require('path');
const { Pool } = require('pg');
const config = require('./config');

const pool = new Pool({ connectionString: config.databaseUrl });

async function runMigrations() {
  const client = await pool.connect();
  try {
    await client.query(`
      CREATE TABLE IF NOT EXISTS schema_migrations (
        version TEXT PRIMARY KEY,
        applied_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
      );
    `);

    const migrationsDir = path.join(__dirname, 'migrations');
    const files = fs.readdirSync(migrationsDir).filter(f => f.endsWith('.sql')).sort();

    const { rows } = await client.query('SELECT version FROM schema_migrations');
    const applied = new Set(rows.map(r => r.version));

    for (const file of files) {
      if (applied.has(file)) {
        console.log(`Skipping applied migration: ${file}`);
        continue;
      }

      console.log(`Applying migration: ${file}...`);
      const sql = fs.readFileSync(path.join(migrationsDir, file), 'utf8');

      await client.query('BEGIN');
      try {
        await client.query(sql);
        await client.query('INSERT INTO schema_migrations (version) VALUES ($1)', [file]);
        await client.query('COMMIT');
        console.log(`Successfully applied: ${file}`);
      } catch (err) {
        await client.query('ROLLBACK');
        console.error(`Failed to apply migration ${file}:`, err.message);
        process.exit(1);
      }
    }

    console.log('All migrations are up to date.');
  } finally {
    client.release();
    await pool.end();
  }
}

runMigrations().catch(err => {
  console.error('Migration runner failed:', err);
  process.exit(1);
});

File: Code/Lesson_7_code/migrate.py
"""Apply every .sql file in migrations/ that has not been applied yet.

Lesson 4 wrote this runner. Lesson 7 changes one thing: the runner now lives
in the container image, and the platform calls it as a release step, before
the new version of the API starts.

    python migrate.py

The runner is idempotent. It reads `schema_migrations`, skips what it finds,
and applies the rest in one transaction each. Run it on every deploy.
"""

import pathlib

import psycopg

from config import settings

MIGRATIONS = pathlib.Path(__file__).parent / "migrations"

with psycopg.connect(settings.dsn) as conn:
    conn.execute("""
        CREATE TABLE IF NOT EXISTS schema_migrations (
            version     TEXT PRIMARY KEY,
            applied_at  TIMESTAMPTZ NOT NULL DEFAULT now()
        )
    """)
    conn.commit()

    rows = conn.execute("SELECT version FROM schema_migrations")
    applied = {v for (v,) in rows}

    for path in sorted(MIGRATIONS.glob("*.sql")):
        version = path.stem
        if version in applied:
            print(f"skip    {version}")
            continue
        with conn.transaction():                     # all-or-nothing
            conn.execute(path.read_text())
            conn.execute("INSERT INTO schema_migrations (version) VALUES (%s)",
                         (version,))
        print(f"applied {version}")
